### PR TITLE
Make manual merge autodetection optional and add manual merge as merge method

### PR DESCRIPTION
**Pull request by a1012112796**
as title, Before this pr, we use same way with GH to check manually merge.It good, but in some special cases, misjudgments can occur. and it's hard to fix this bug. So I add option to allow repo manager block "auto check manual merge" function, Then it will have same style like gitlab(allow empty pr). and to compensate for not being able to detect the pr merge automatically, I added a manual approach.

example view:
tmpe

### CODE REVIEW

File: integrations/pull_status_test.go [hunk 1]
```diff
@@ -114,7 +114,7 @@ func TestPullCreate_EmptyChangesWithCommits(t *testing.T) {
 		resp := session.MakeRequest(t, req, http.StatusOK)
 		doc := NewHTMLParser(t, resp.Body)
 
-		text := strings.TrimSpace(doc.doc.Find(".item.text.green").Text())
-		assert.EqualValues(t, "This pull request can be merged automatically.", text)
+		text := strings.TrimSpace(doc.doc.Find(".item.text.grey").Text())
+		assert.EqualValues(t, "This branche is equal with target branch.", text)
```
Comment (zeripath) [inline integrations/pull_status_test.go]:
```suggestion
		assert.EqualValues(t, "This branch is equal with the target branch.", text)
```

File: options/locale/locale_en-US.ini [hunk 2]
```diff
@@ -1176,13 +1178,16 @@ pulls.reopen_to_merge = Please reopen this pull request to perform a merge.
 pulls.cant_reopen_deleted_branch = This pull request cannot be reopened because the branch was deleted.
 pulls.merged = Merged
 pulls.merged_as = The pull request has been merged as <a rel="nofollow" class="ui sha" href="%[1]s"><code>%[2]s</code></a>.
+pulls.manually_merged = Merged manually
+pulls.manually_merged_as = The pull request has been merged as <a rel="nofollow" class="ui sha" href="%[1]s"><code>%[2]s</code></a> manually.
```
Comment (zeripath) [inline options/locale/locale_en-US.ini]:
```suggestion
pulls.manually_merged_as = The pull request has been manually merged as <a rel="nofollow" class="ui sha" href="%[1]s"><code>%[2]s</code></a>.
```

File: options/locale/locale_en-US.ini [hunk 3]
```diff
@@ -1176,13 +1178,16 @@ pulls.reopen_to_merge = Please reopen this pull request to perform a merge.
 pulls.cant_reopen_deleted_branch = This pull request cannot be reopened because the branch was deleted.
 pulls.merged = Merged
 pulls.merged_as = The pull request has been merged as <a rel="nofollow" class="ui sha" href="%[1]s"><code>%[2]s</code></a>.
+pulls.manually_merged = Merged manually
```
Comment (zeripath) [inline options/locale/locale_en-US.ini]: I think I prefer:

```suggestion
pulls.manually_merged = Manually merged
```

File: options/locale/locale_en-US.ini [hunk 4]
```diff
@@ -1176,13 +1178,16 @@ pulls.reopen_to_merge = Please reopen this pull request to perform a merge.
...
+pulls.manually_merged_as = The pull request has been merged as <a rel="nofollow" class="ui sha" href="%[1]s"><code>%[2]s</code></a> manually.
 pulls.is_closed = The pull request has been closed.
 pulls.has_merged = The pull request has been merged.
 pulls.title_wip_desc = `<a href="#">Start the title with <strong>%s</strong></a> to prevent the pull request from being merged accidentally.`
 pulls.cannot_merge_work_in_progress = This pull request is marked as a work in progress. Remove the <strong>%s</strong> prefix from the title when it's ready
 pulls.data_broken = This pull request is broken due to missing fork information.
 pulls.files_conflicted = This pull request has changes conflicting with the target branch.
 pulls.is_checking = "Merge conflict checking is in progress. Try again in few moments."
+pulls.is_empty = "This branche is equal with target branch."
```
Comment (zeripath) [inline options/locale/locale_en-US.ini]:
```suggestion
pulls.is_empty = "This branch is equal with the target branch."
```

Are we sure this should be `target` and not `base`?

```suggestion
pulls.is_empty = "This branch is equal with base branch."
```

File: options/locale/locale_en-US.ini [hunk 5]
```diff
@@ -1200,6 +1205,7 @@ pulls.reject_count_1 = "%d change request"
 pulls.reject_count_n = "%d change requests"
 pulls.waiting_count_1 = "%d waiting review"
 pulls.waiting_count_n = "%d waiting reviews"
+pulls.wrong_commit_id = "commit id is wrong, need real commit id on base branch"
```
Comment (zeripath) [inline options/locale/locale_en-US.ini]:
```suggestion
pulls.wrong_commit_id = "commit id must be a commit id on the base branch"
```

**Review by zeripath — COMMENTED** [review]
Cool! I looked at keeping the automatic merge and just selecting force pushes to switch this off but actually this is much simpler. 

I need to check your code but first of all some quick labelling comments:

File: models/repo_unit.go [hunk 6]
```diff
@@ -93,6 +93,8 @@ type PullRequestsConfig struct {
 	AllowRebase               bool
 	AllowRebaseMerge          bool
 	AllowSquash               bool
+	AllowManualMerge          bool
+	AutoCheckManualMerge      bool
```
Comment (zeripath) [inline models/repo_unit.go]: Why not make this `AutomaticManualMerge`

File: routers/api/v1/repo/pull.go [hunk 7]
```diff
@@ -759,13 +759,70 @@ func MergePullRequest(ctx *context.APIContext, form auth.MergePullRequestForm) {
...
+			log.Error("PullRequest[%d].setMerged : %v", pr.ID, err)
+			ctx.Status(500)
+			return
+		} else if !merged {
+			return
+		}
+
+		notification.NotifyMergePullRequest(pr, ctx.User)
+		log.Info("manuallyMerged[%d]: Marked as manually merged into %s/%s by commit id: %s", pr.ID, pr.BaseRepo.Name, pr.BaseBranch, commit.ID.String())
+		ctx.Status(http.StatusOK)
```
Comment (zeripath) [inline routers/api/v1/repo/pull.go]: This belongs in a function in services/pull then you wouldn't have to duplicate it routers/repo/pull.go

File: routers/api/v1/repo/pull.go [hunk 8]
```diff
@@ -759,13 +759,70 @@ func MergePullRequest(ctx *context.APIContext, form auth.MergePullRequestForm) {
...
+		branchName, err := commit.GetBranchName()
+		if err != nil {
+			ctx.ServerError("commit.GetBranchName", err)
+			return
+		}
+
+		if branchName != pr.BaseBranch {
+			ctx.Error(http.StatusMethodNotAllowed, "commit id is wrong, need real commit id on base branch", "")
+			return
+		}
```
Comment (zeripath) [inline routers/api/v1/repo/pull.go]: So this is unnecessarily strict I think.

You could reasonably argue that any commit on the base branch is a reasonable place to say this was merged to.

There are several ways to detect that:

```
git branch --contains COMMIT_ID BRANCH_NAME
```

where if it returns anything then the commit is on the branch

or:

```
git rev-list BRANCH_NAME | grep COMMIT_ID
```

would also work.

**Review by zeripath — APPROVED** [review]

**Comment by zeripath** [comment]
I've fixed the tests and improved the reporting of manual merge not allowed in the API

**Comment by a1012112796** [comment]
@zeripath Thanks for your help. I found these codes are marked as "red" in the codecove report, I wonder the reason, Maybe it still have some bug? 

image

link: https://codecov.io/gh/go-gitea/gitea/src/5d1b12c1ba4e462aeefe1edd5b388b0d3a4076c4/services/pull/merge.go

**Comment by a1012112796** [comment]
Need more work about test code :(

**Review by zeripath — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Comment by techknowlogick** [comment]
🚀

**Pull request merged**